### PR TITLE
Implement Tag expressions

### PR DESCRIPTION
**Pull request by zdenek-jonas**
Hi, i have implement the current Cucumber tag expression. In this version cucumber Still supports both variants of tags. So i can totally remove the TagFilter. Just now it works by the same way as Cucumber. I have increase the Version of gherkin, so i have to rewrite some part of code. Thanks to use Cucumber itself, we do not need to test the Tag filtering more. It just now part of Cucumber projects. 

The groovy test based on the filepath with "classpath". I have no time to fix these old tests. And this tests will be very quickly old, because the old tag behaviour is now deprecated. 

Please give me a feedback to this code. I am open for it. 
have a nice day
Zdenek Jonas

### CODE REVIEW

**Comment by johny2000uwb** [comment]
Sorry, I will need a more explanation what i should do. I use Cucumber-core. There is a Class TagPredicate. Should i copy this classes (functionality) to this plugin? Or is this way totally bad?

**Comment by mpkorstanje** [comment]
Copy it. And the tests.

**Comment by johny2000uwb** [comment]
I will add the tests, but this i need to commit and i do not want to lose this thread.

**Comment by mpkorstanje** [comment]
Wow. Bit of a misunderstanding here. 

I was only talking about the implementation of `TagPredicate`. Not the whole tag expression library. You can add that as a dependency.

**Comment by johny2000uwb** [comment]
Ok... i am going to remove it....

**Comment by johny2000uwb** [comment]
so now?

**Comment by mpkorstanje** [comment]
Looks much better! Can you remove the whitespace only changes and squash your commits? This is easy to do if you reset your branch and use a diff viewer to revert the chunks that are only whitespace. 

Removing the whitespace changes makes it easier for people to see what you've actually changed. The squash makes the history a bit more understandable.

File: src/it/junit/filter-tag-expressions-scenario/pom.xml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                <executions>
+                    <execution>
+                        <id>generateRunners</id>
+                        <phase>generate-test-sources</phase>
+                        <goals>
+                            <goal>generateRunners</goal>
+                        </goals>
+                        <configuration>
+                        	<tags>
+                        		<tag>@test_for_england or not (@!test_for_england)</tag>
```
Comment (mpkorstanje) [inline src/it/junit/filter-tag-expressions-scenario/pom.xml]: I had to puzzle on this for a minute. Can you come up with another tag name for `@!test_for_england`.

File: src/it/junit/filter-tag-expressions-scenario/src/test/resources/features/feature1.feature [hunk 2]
```diff
@@ -0,0 +1,16 @@
+Feature: Feature1
+@test_for_england
+
+  Scenario: Generate Junit Runner for each feature file
```
Comment (mpkorstanje) [inline src/it/junit/filter-tag-expressions-scenario/src/test/resources/features/feature1.feature]: The tag should be right above the scenario line.

File: src/it/junit/filter-tag-expressions-scenario/pom.xml (hunk 1 above)
Comment (mpkorstanje) [reply src/it/junit/filter-tag-expressions-scenario/pom.xml]: Same for the other integration test

File: src/it/junit/filter-tag-expressions-scenario/src/test/resources/features/feature1.feature (hunk 2 above)
Comment (mpkorstanje) [reply src/it/junit/filter-tag-expressions-scenario/src/test/resources/features/feature1.feature]: Same for the other integration test

File: src/it/junit/filter-tag-expressions-scenario/pom.xml (hunk 1 above)
Comment (johny2000uwb) [reply src/it/junit/filter-tag-expressions-scenario/pom.xml]: ok, @not_for_england...  will be better

File: src/it/junit/filter-tag-expressions-scenario/src/test/resources/features/feature1.feature (hunk 2 above)
Comment (johny2000uwb) [reply src/it/junit/filter-tag-expressions-scenario/src/test/resources/features/feature1.feature]: true... i will fix it

File: src/it/junit/filter-tag-expressions-scenario/pom.xml (hunk 1 above)
Comment (johny2000uwb) [reply src/it/junit/filter-tag-expressions-scenario/pom.xml]: Europe and Germany much better

Comment (mpkorstanje) [reply src/it/junit/filter-tag-expressions-scenario/pom.xml]: Heh.  https://www.youtube.com/watch?v=NAE1YBxEE38

**Comment by johny2000uwb** [comment]
Ok, whitespaces are out. But with the "squash" i have no experience. I have to study how to do it. Hopefully i will find something for eclipse.

**Comment by mpkorstanje** [comment]
https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

**Comment by johny2000uwb** [comment]
Ok, Squash and whitespaces finished.

**Review by mpkorstanje — APPROVED** [review]

**Comment by mpkorstanje** [comment]
@temyers I've set things up for a new release (major version bump, migration notes, ect). Can you build it?

**Comment by johny2000uwb** [comment]
Could you please release it, i need that.

**Comment by mpkorstanje** [comment]
It's unfunded open source. It moves at the speed of peoples spare time. You can always deploy your own version to your local nexus though. Just change the groupId to avoid confusion.

**Comment by temyers** [comment]
5.0.0 released.

It should be available on Maven Central shortly.

**Comment by johny2000uwb** [comment]
@temyers it is already in Mavenrepository. Thank you.
@mpkorstanje Thank you for feedback and help with the implementation. It was my first open source pull request.

**Comment by mpkorstanje** [comment]
You're welcome!

**Pull request merged**